Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift

```diff
@@ -159,40 +166,51 @@ public final class ProjectDescriptionHelpersBuilder: ProjectDescriptionHelpersBu
         projectDescriptionSearchPaths: ProjectDescriptionSearchPaths,
         customProjectDescriptionHelperModules: [ProjectDescriptionHelpersModule] = []
     ) async throws -> ProjectDescriptionHelpersModule {
-        if let cachedModule = builtHelpers.withValue({ $0[path] }) { return cachedModule }
-
-        let hash = try projectDescriptionHelpersHasher.hash(helpersDirectory: path)
-        let prefixHash = projectDescriptionHelpersHasher.prefixHash(helpersDirectory: path)
+        let projectDescriptionHelpersModuleBuild = try builtHelpers.mutate { cache in
+            if let cachedModule = cache[path] {
+                return cachedModule
+            }
 
-        let helpersCachePath = cacheDirectory.appending(component: prefixHash)
-        let helpersModuleCachePath = helpersCachePath.appending(component: hash)
-        let dylibName = "lib\(name).dylib"
-        let modulePath = helpersModuleCachePath.appending(component: dylibName)
-        let projectDescriptionHelpersModule = ProjectDescriptionHelpersModule(name: name, path: modulePath)
+            let hash = try projectDescriptionHelpersHasher.hash(helpersDirectory: path)
+            let moduleCacheDirectory = cacheDirectory.appending(component: hash)
+            let dylibName = "lib\(name).dylib"
+            let modulePath = moduleCacheDirectory.appending(component: dylibName)
 
-        builtHelpers.mutate { $0[path] = projectDescriptionHelpersModule }
+            let module = ThreadSafe(
+                HelpersModuleBuild(
+                    module: ProjectDescriptionHelpersModule(name: name, path: modulePath)
+                )
+            )
+            cache[path] = module
 
-        if FileHandler.shared.exists(helpersModuleCachePath) {
-            return projectDescriptionHelpersModule
+            return module
         }
 
-        try FileHandler.shared.createFolder(helpersModuleCachePath)
+        return try projectDescriptionHelpersModuleBuild.mutate { build in
+            if build.isBuilt, FileHandler.shared.exists(build.module.path) {
```

Comment:
I don't know how likely it is to happen, but if helpers are currently being built, doesn't this mean we'll rebuild them again, unnecessarily? Should we instead save the project description helpers builder process using something like `Task` and wait for the result instead of starting the process again?

Reply:
> doesn't this mean we'll rebuild them again, unnecessarily? 
> starting the process again

You are right. This is possible if helper builder is initialized again.
And the reason is the `build.isBuilt` flag which is always false by default.

1. We can simply remove this flag:
I added it because... suggested that the module file might be created during the build process rather than when the command completes.

2. We can create a lock file in this directory to block re-building.
But I don’t yet know how to implement the process of waiting for the lock file to be deleted

Reply:
UPD: replace on Task